README.md
# AI_Draw_safemode
重构代码结构（屎山自己都看不下去了），暂时这样用先，预计增加TAG筛选显示功能
## 功能介绍
通过上传图片参数至数据库，来实现快捷的图片预览、图片TAG获取、对应标签的ai图片生成等功能。
## 具体操作
1、下载aidraw_db.py，放到<code>hoshino/modules/aidrawDB</code>文件夹中\
2、在<code>hoshino/config/__bot__.py</code>文件中，<code>MODULES_ON</code>里添加 "aidrawDB"\
3、<b>打开aidraw_db.py，在17行填入ai绘图的API接口，在18行填入个人token</b>
## 功能指令
|  指令   | 说明  | 具体  |
|  ----  | ----  | ----  |
| <b>绘图数据库帮助</b>  |  查看该功能的全部指令  | 以图片形式生成  |
| 上传<b>[参数]</b>  | 上传图片和tag至数据库内  | <b>[参数]</b>为ai绘图的指令  |
| <b>'回复'</b>[bot昵称]上传  | 通过回复他人的ai绘图来进行上传  | 具体下图教程  |
| 炼金手册<b>[序号]</b>  | 查询炼金手册内容  | <b>[序号]</b>为页数，每页显示8张  |
| 查看配方<b>[序号]</b>  | 查询炼金配方内容  | <b>[序号]</b>为图片标签  |
| 使用配方<b>[序号]</b>  | 使用炼金配方内容  | <b>[序号]</b>为图片标签  |
| 删除配方<b>[序号]</b>  | 删除炼金配方内容  | <b>[序号]</b>为图片标签  |
PS:上传仅限管理员使用，删除仅限超级管理员使用
## 演示
<b>绘图数据库帮助:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E7%BB%98%E5%9B%BE%E6%95%B0%E6%8D%AE%E5%BA%93%E5%B8%AE%E5%8A%A9.png)
<b>上传[参数]:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E4%B8%8A%E4%BC%A0.png)
<b>'回复'[bot昵称]上传:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E6%8C%87%E4%BB%A4%E4%B8%8A%E4%BC%A0.png)
<b>炼金手册[序号]:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E7%82%BC%E9%87%91%E6%89%8B%E5%86%8C.png)
<b>查看配方[序号]:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E6%9F%A5%E7%9C%8B%E9%85%8D%E6%96%B9.png)
<b>使用配方[序号]:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E4%BD%BF%E7%94%A8%E9%85%8D%E6%96%B9.png)
<b>删除配方[序号]:</b>
![image](https://github.com/jiyemengmei/AI_Draw_safemode/blob/main/image/%E5%88%A0%E9%99%A4%E9%85%8D%E6%96%B9.png)
## 鸣谢
<a href="https://github.com/Mrs4s/go-cqhttp" target="_BLANK">go-cqhttp</a>\
<a href="https://github.com/Ice-Cirno/HoshinoBot" target="_BLANK">HoshinoBot</a>\
<a href="" target="_BLANK">Cath的AI绘图源码</a>
## API
<a href="" target="_BLANK">路路佬私人API</a>
## 更新日志
### 2022/10/9
首次上传


CHANGELOG.md
# CHANGELOG
- 2022-10-11: 
  - 缝合 XP 查询, 以图生图, 正常文字转图片, 添加配置文件, 合并转发消息(回复上传未完成)
    - SIMSUNB.TTF 字体文件可以从如下链接获取: https://www.aliyundrive.com/s/SkeZxnbrqgB

xp.py
# XP 查询 by 季落
import os
import sqlite3

# XP 数据库路径
XP_DB_PATH = os.path.expanduser('~/.hoshino/AI_image_xp.db')
class XpCounter:
    def __init__(self):
        os.makedirs(os.path.dirname(XP_DB_PATH), exist_ok=True)
        self._create_table()
    def _connect(self):
        return sqlite3.connect(XP_DB_PATH)
        
    def _create_table(self):
        try:
            self._connect().execute(
                    '''CREATE TABLE IF NOT EXISTS XP_NUM
                        (UID             INT    NOT NULL,
                        KEYWORD         TEXT   NOT NULL,
                        NUM             INT    NOT NULL,
                        PRIMARY KEY(UID,KEYWORD));
                    '''
            )
        except:
            raise Exception('创建表发生错误')
            
    def _add_xp_num(self, uid, keyword):
        try:
        
            num = self._get_xp_num(uid, keyword)
            if num == None:
                num = 0
            num += 1
            with self._connect() as conn:
                conn.execute(
                    "INSERT OR REPLACE INTO XP_NUM (UID,KEYWORD,NUM) \
                                VALUES (?,?,?)", (uid, keyword, num)
                )
        except:
            raise Exception('更新表发生错误')
            
    def _get_xp_num(self, uid, keyword):
        try:
            r = self._connect().execute("SELECT NUM FROM XP_NUM WHERE UID=? AND KEYWORD=?", (uid, keyword)).fetchone()
            return 0 if r is None else r[0]
        except:
            raise Exception('查找表发生错误')
    
    def _get_xp_list(self, uid, num):
        """获取个人 xp 列表"""
        with self._connect() as conn:
            r = conn.execute(
                f"SELECT KEYWORD,NUM FROM XP_NUM WHERE UID={uid} ORDER BY NUM desc LIMIT {num}").fetchall()
        return r if r else {}

    def _get_xp_list_all(self, num):
        """获取全体 xp 列表前 num 项"""
        with self._connect() as conn:
            r = conn.execute(
                f"SELECT KEYWORD,NUM FROM XP_NUM ORDER BY NUM desc LIMIT {num}").fetchall()
        return r if r else {}

def add_xp_num(uid, keyword):
    """xp 计数"""
    XP = XpCounter()
    XP._add_xp_num(uid, keyword)

def get_xp_list(uid):
    """获取个人 xp 列表(前15项"""
    XP = XpCounter()
    xp_list = XP._get_xp_list(uid, 15)
    if len(xp_list)>0:
        data = sorted(xp_list,key=lambda cus:cus[1],reverse=True)
        new_data = []
        for xp_data in data:
            keyword, num = xp_data
            new_data.append((keyword,num))
        rankData = sorted(new_data,key=lambda cus:cus[1],reverse=True)
        # print(f'返回查询结果为: {rankData}')
        return rankData
    else:
        return []


def get_xp_list_all():
    """获取群友的 xp 列表(前15项)"""
    xp = XpCounter()
    # print('已进入查询函数, 准备调用数据库查询函数查询')
    xp_list = xp._get_xp_list_all(15)
    # print(f'''查询结果为: {xp_list}''')
    if len(xp_list)>0:
        data = sorted(xp_list,key=lambda cus:cus[1],reverse=True)
        new_data = []
        for xp_data in data:
            keyword, num = xp_data
            new_data.append((keyword,num))
        rankData = sorted(new_data,key=lambda cus:cus[1],reverse=True)
        return rankData
    else:
        return []


def add_xp_num(uid, keyword):
    XP = XpCounter()
    XP._add_xp_num(uid, keyword)



alchemy_manual.py
# 炼金手册数据库操作
from os.path import dirname, join, exists
from os import makedirs, remove
import sqlite3
from hoshino.typing import CQEvent
import requests
from PIL import Image, ImageDraw,ImageFont
import re
from io import BytesIO
import base64
import calendar, time
from hoshino.config import NICKNAME
from aiocqhttp.exceptions import ActionFailed


curpath = dirname(__file__)
image_list_db = join(curpath, 'save_tags.db')   #保存tags数据库
save_image_path= join(curpath,'SaveImage')  # 保存图片路径

class AlchemyManual:
    def __init__(self):
        makedirs(dirname(image_list_db), exist_ok=True)
        makedirs(save_image_path, exist_ok=True)
        self._create_table()
    
    def _connect(self):
        return sqlite3.connect(image_list_db)

    def _create_table(self):
        try:
            self._connect().execute(
                """CREATE TABLE IF NOT EXISTS aitag(
                    scale TINYINT,
                    size CHAR,
                    tags TEXT,
                    seed INT,
                    saveconfig BLOB
                )
                """
            )
        except:
            raise Exception('创建炼金手册tags数据库发生错误')
    
    def _upload_recipe(self, scale, size, tags, seed, saveconfig):
        try:
            with self._connect() as conn:
                conn.execute(
                    "INSERT INTO aitag (scale, size, tags, seed, saveconfig) VALUES (?,?,?,?,?)", (scale, size, tags, seed, saveconfig)
                )
        except:
            raise Exception('上传配方时发生错误')

    def _get_alchemy_manual(self):
        """获取炼金手册, 返回由 (rowid, saveconfig) 组成的列表"""
        try:
            with self._connect() as conn:
                return conn.execute("SELECT rowid, saveconfig FROM aitag").fetchall()
        except:
            raise Exception('获取配方时发生错误')

    def _get_recipe(self, rowid):
        """获取配方, 返回所有信息, rowid: 配方id"""
        try:
            with self._connect() as conn:
                return conn.execute("SELECT * FROM aitag WHERE rowid=?", (rowid,)).fetchone()
        except:
            raise Exception('获取配方时发生错误')
    
    def _delete_recipe(self, rowid):
        """删除配方, rowid: 配方id"""
        try:
            with self._connect() as conn:
                conn.execute("DELETE FROM aitag WHERE rowid=?", (rowid,))
        except:
            raise Exception('删除配方时发生错误')
    

def upload_recipe(ev: CQEvent) -> tuple:
    """上传配方
    
    Args:
        ev (CQEvent): event

    Returns: tuple(code, str)
        code - 0: 失败 1: 成功
        str:上传状态, 包括:
            上传成功
            图片格式出错
            种子格式出错
            权重格式出错
            标签格式出错
            其他报错
    """
    alchemy_manual = AlchemyManual()
    try:
        for i in ev.message:
            if i.type == "image":
                image=str(i)
                break
        image_url = re.match(r"\[CQ:image,file=(.*),url=(.*)\]", str(image))
        pic_url = image_url.group(2)
        response = requests.get(pic_url)
        img = Image.open(BytesIO(response.content)).convert("RGB")
        ls_f=base64.b64encode(BytesIO(response.content).read())
        imgdata=base64.b64decode(ls_f)
        datetime = calendar.timegm(time.gmtime())
        image_path= './SaveImage/'+str(datetime)+'.png'
        saveconfig = join(curpath, f'{image_path}')
        size=f"{img.width}x{img.height}"
        pic = open(saveconfig, "wb")
        pic.write(imgdata)
        pic.close()
    except:
        return (0, '图片格式出错')
    try:
        seed_list1=str(ev.message).split(f"scale:")
        seed_list2=seed_list1[0].split('eed:')
        seed=seed_list2[1].strip ()
    except:
        return (0, '种子格式出错')
    try:
        scale_list=seed_list1[1].split(f"tags:")
        scale=scale_list[0].strip()
    except:
        return (0, '权重格式出错')
    try:
        tags=scale_list[1].strip()
    except:
        return (0, '标签格式出错')
    try:
        alchemy_manual._upload_recipe(scale, size, tags, seed, saveconfig)
        return (1, '上传成功')
    except Exception as e:
        return (0, f'其他报错: {e}')


def upload_recipe_by_reply(real_message) -> tuple:
    """通过回复 bot 消息上传配方"""
    try:
        # print(f'数据库查询函数接收到的消息: {real_message}')
        image_url = re.search(r"\[CQ:image,file=(.*),url=(.*)\]", real_message)
        # print(f'image_url为: {image_url}')
        if not image_url:
            return (0, '未找到图片')
        file = image_url.group(1)
        pic_url = image_url.group(2)
        if ',subType=' in pic_url:
            sbtype=pic_url.split('=')[-1]
            pic_url = pic_url.split(',')[0]
        elif ',subType=' in file:
            sbtype=file.split('=')[-1]
            file = file.split(',')[0]
        else:
            sbtype=None
        if 'c2cpicdw.qpic.cn/offpic_new/' in pic_url:
            md5 = file[:-6].upper()
            pic_url = f"http://gchat.qpic.cn/gchatpic_new/0/0-0-{md5}/0?term=2"
        response = requests.get(pic_url)
        img = Image.open(BytesIO(response.content)).convert("RGB")
        ls_f=base64.b64encode(BytesIO(response.content).read())
        imgdata=base64.b64decode(ls_f)
        datetime = calendar.timegm(time.gmtime())
        image_path= './SaveImage/'+str(datetime)+'.png'
        saveconfig = join(curpath, f'{image_path}')
        size=f"{img.width}x{img.height}"
        pic = open(saveconfig, "wb")
        pic.write(imgdata)
        pic.close()
    except:
        return (0, '图片格式出错')
    try:
        # 匹配传入字符串中 seed - scale 的部分并去除空格也即 seed 的值
        seed_prefix = 'seed'
        seed_suffix = 'scale'
        seed = real_message[real_message.find(seed_prefix) + len(seed_prefix):real_message.find(seed_suffix)].strip()
    except:
        return (0, '种子格式出错')
    try:
        # 匹配传入字符串中 scale - tags 的部分并去除空格也即 scale 的值
        scale_prefix = 'scale'
        scale_suffix = 'tags'
        scale = real_message[real_message.find(scale_prefix) + len(scale_prefix):real_message.find(scale_suffix)].strip()
    except:
        return (0, '权重格式出错')
    try:
        # 匹配传入字符串中 tags - 末尾 的部分并去除空格也即 tags 的值
        tags_prefix = 'tags'
        tags = real_message[real_message.find(tags_prefix) + len(tags_prefix):].strip()
    except:
        return (0, '标签格式出错')
    try:
        conn=sqlite3.connect(image_list_db)
        cur = conn.cursor()
        cur.execute("INSERT INTO aitag VALUES (?,?,?,?,?)",(scale,size,tags,seed,saveconfig))
        conn.commit()
        cur.close()
        conn.close()
        return (1, '上传成功!')
    except Exception as e:
        return (0, f'其他报错: {e}')

ttf_font_path = join(dirname(__file__), 'SIMSUNB.TTF')


def get_alchemy_manual(page: int = 1):
    """获取配方 page:页码, 默认为1"""
    alchemy_manual = AlchemyManual()
    image_list = alchemy_manual._get_alchemy_manual()
    target = Image.new('RGB', (1920,1080),(255,255,255))
    i=0
    for index in range(0+(page*8),8+(page*8)):
        try:
            image_msg=image_list[index]
        except:
            break
        rowid = image_msg[0]
        image_path=image_msg[1]
        region = Image.open(image_path)
        region = region.convert("RGB")
        region = region.resize((int(region.width/2),int(region.height/2)))
        font = ImageFont.truetype(ttf_font_path, 36)  # 设置字体和大小
        draw = ImageDraw.Draw(target)
        if i<4:
            target.paste(region,(80*(i+1)+384*i,50))
            draw.text((80*(i+1)+384*i+int(region.width/2)-18,80+region.height),str(rowid).replace(',',''),font=font,fill = (0, 0, 0))
        if i>=4:
            target.paste(region,(80*(i-3)+384*(i-4),150+384))
            draw.text((80*(i-3)+384*(i-4)+int(region.width/2)-18,180+384+region.height),str(rowid).replace(',',''),font=font,fill = (0, 0, 0))
        i+=1
    result_buffer = BytesIO()
    target.save(result_buffer, format='JPEG', quality=100) #质量影响图片大小
    imgmes = 'base64://' + base64.b64encode(result_buffer.getvalue()).decode()
    resultmes = f"[CQ:image,file={imgmes}]"
    return (1, resultmes)


def get_recipe(rowid: int):
    """获取配方 rowid:配方id"""
    alchemy_manual = AlchemyManual()
    recipe = alchemy_manual._get_recipe(rowid)
    scale, size, tags, seed, image_path = recipe

    pic = open(image_path, "rb")
    base64_str = base64.b64encode(pic.read())
    imgmes = 'base64://' + base64_str.decode()
    resultmes = f"[CQ:image,file={imgmes}]"
    pic.close()
    msg=f"序号为:{rowid}\n{resultmes}\n配方:ai绘图{tags}\nCFG scale: {scale}, Size:{size}"
    return (1, msg)


def use_recipe(rowid: int):
    """使用配方 rowid:配方id"""
    alchemy_manual = AlchemyManual()
    recipe = alchemy_manual._get_recipe(rowid)
    scale, size, tags, seed, image_path = recipe

    msg=f"{tags}\nCFG scale: {scale}, Size:{size}".replace('&r18=1','')
    return (1, msg)


def delete_recipe_by_rowid(rowid: int):
    """删除配方 rowid:配方id"""
    alchemy_manual = AlchemyManual()
    recipe = alchemy_manual._get_recipe(rowid)
    scale, size, tags, seed, image_path = recipe
    remove(image_path)
    alchemy_manual._delete_recipe(rowid)
    return (1, f'已删除配方{rowid}')



aidraw_db.py
from asyncio import Lock
from os.path import dirname, join, exists
from urllib import response
from hoshino import Service,priv,aiorequests
from hoshino.util import FreqLimiter, DailyNumberLimiter
from hoshino.config import NICKNAME
from hoshino.typing import CQEvent
from aiocqhttp.exceptions import ActionFailed
from PIL import Image, ImageDraw,ImageFont
from io import BytesIO
import requests


import base64
import json
import time,calendar
import re
import sqlite3
import os
import toml

from . import xp
from . import alchemy_manual

# 导入配置文件
toml_config_path = join(dirname(__file__), 'config.toml')
config = toml.load(toml_config_path)
word2img_url = config['word2img_url']   # 文字 -> 图片 apiURL
img2img_url = config['img2img_url']     # 以图生图 apiURL
token = config['token'] # token
# 口令间隔时间(默认每 60 秒可以使用一次口令)
interval = FreqLimiter(config['interval'] )
# 每日口令使用次数(默认每天可以使用 20 次口令)
daily_limit_ = config['daily_limit']
daily_limit = DailyNumberLimiter(daily_limit_)


sv_help = '''
【AI绘图数据库】
-[上传[参数]]  上传图片和tag至数据库内, [参数]为ai绘图的指令
-[炼金手册[序号]]  查询炼金手册内容，[序号]为页数, 每页显示8张
-[查看配方[序号]]  查询炼金配方内容，[序号]为图片标签
-[使用配方[序号]]  使用炼金配方内容，[序号]为图片标签
-[删除配方[序号]]  删除炼金配方内容，[序号]为图片标签
PS:上传仅限管理员使用，删除仅限超级管理员使用
'''

sv = Service(
    name = 'AI绘图数据库',  #功能名
    use_priv = priv.NORMAL, #使用权限   
    manage_priv = priv.ADMIN, #管理权限
    visible = True, #是否可见
    enable_on_default = True, #是否默认启用
    bundle = '娱乐', #属于哪一类
    help_ = sv_help #帮助文本
    )

ttf_font_path = join(dirname(__file__), 'SIMSUNB.TTF')

async def wordimage(word):
    bg = Image.new('RGB', (950,300), color=(255,255,255))
    font = ImageFont.truetype(ttf_font_path, 30)  # 设置字体和大小
    draw = ImageDraw.Draw(bg)
    draw.text((10,5), word, fill="#000000", font=font)
    result_buffer = BytesIO()
    bg.save(result_buffer, format='JPEG', quality=100)
    imgmes = 'base64://' + base64.b64encode(result_buffer.getvalue()).decode()
    resultmes = f"[CQ:image,file={imgmes}]"
    return resultmes

@sv.on_fullmatch(["绘图数据库帮助"])
async def bangzhu(bot, ev):
    await bot.send(ev, await wordimage(sv_help), at_sender=True)



@sv.on_fullmatch(['我的XP'])
async def get_my_xp(bot, ev: CQEvent):
    xp_list = xp.get_xp_list(ev.user_id)
    # uid = ev.user_id
    msg = '您的XP信息为:\n'
    if len(xp_list)>0:
        for xpinfo in xp_list:
            keyword, num = xpinfo
            msg += f'关键词：{keyword}；查询次数：{num}\n'
    else:
        msg += '暂无您的XP信息'
    await bot.send(ev, msg, at_sender=True)


@sv.on_fullmatch(['群友的XP'])
async def get_group_xp(bot, ev: CQEvent):
    xp_list = xp.get_xp_list_all()
    msg = '群友的XP信息为:\n'
    if len(xp_list)>0:
        for xpinfo in xp_list:
            keyword, num = xpinfo
            msg += f'关键词：{keyword}；查询次数：{num}\n'
    else:
        msg += '暂无群友的XP信息'
    await bot.send(ev, msg, at_sender=True)



@sv.on_prefix(('上传'))
async def upload_header(bot, ev):
    """上传配方"""
    if not priv.check_priv(ev, priv.ADMIN):
        await bot.finish(ev, '上传配方仅限管理员使用', at_sender=True)
        return
    response = alchemy_manual.upload_recipe(ev)
    if response[0] == 0:
        await bot.finish(ev, response[1], at_sender=True)
    else:
        await bot.send(ev, response[1], at_sender=True)
        

@sv.on_message('group')
async def replymessage(bot, ev: CQEvent):
    """通过回复 bot 的合并转发消息上传"""
    print(f'收到来自{ev.user_id}的消息：{ev.message}, 类型：{ev.message_type}')
    seg=ev.message[0]
    if seg.type != 'reply':
        return
    tmid = seg.data['id']
    cmd = ev.message.extract_plain_text()
    flag1 = 0
    flag2 = 0
    for m in ev.message[2:]:
        if m.type == 'at' and m.data['qq'] == ev.self_id:
            flag1 = 1
    for name in NICKNAME:
        if name in cmd:
            flag1 = 1
            break
    for pfcmd in ['上传', '窃取', '偷了']:
        if pfcmd in cmd:
            flag2 = 1
    if not (flag1 and flag2):
        return
    if not priv.check_priv(ev, priv.ADMIN):
        await bot.finish(ev,f"仅限管理员上传！", at_sender=True)
        return
    try:
        tmsg = await bot.get_msg(self_id=ev.self_id, message_id=int(tmid))
    except ActionFailed:
        await bot.finish(ev, '该消息已过期，请重新转发~')
        return
    print(f'获取到的消息为: {tmsg}')
    # 获取合并抓发消息的 id(前 15 个字符为 [CQ:forward,id=])
    forward_id = tmsg['message'][15:-1]
    print(f'获取到的合并转发消息id为: {forward_id}')
    # 获取转发消息中的第一条消息
    print('正在获取转发消息中的第一条消息')
    message_forward = await bot.get_forward_msg(id=forward_id)
    print(f'获取到的转发消息为: {message_forward}, 类型为: {type(message_forward)}')
    real_message = message_forward['messages'][0]['content']
    print(f'获取到的转发消息中的第一条消息为: {real_message}')

    response = alchemy_manual.upload_recipe_by_reply(real_message)
    if response[0] == 0:
        await bot.finish(ev, response[1], at_sender=True)
    else:
        await bot.send(ev, response[1], at_sender=True)



@sv.on_rex((r'^炼金手册([1-9]\d*)$'))
async def alchemy_book(bot, ev):
    match = ev['match']
    page=int(match.group(1))-1
    response = alchemy_manual.get_alchemy_manual(page)
    if response[0] == 1:
        await bot.send(ev, response[1])


@sv.on_rex((r'^查看配方([1-9]\d*)'))
async def view_recipe(bot, ev):
    uid = str(ev['user_id'])
    if not interval.check(uid):
        await bot.finish(ev, f'魔力回复中！(剩余 {int(interval.left_time(uid)) + 1}秒)', at_sender=True)
        return
    interval.start_cd(uid,30)
    match = ev['match']
    rowid=int(match.group(1))

    response = alchemy_manual.get_recipe(rowid)
    if response[0] == 1:
        # print(f'发送配方消息：{response[1]}')
        await bot.send(ev, response[1])


@sv.on_rex((r'^使用配方([1-9]\d*)'))
async def generate_recipe(bot, ev):
    uid = str(ev['user_id'])
    if not daily_limit.check(uid):
        await bot.finish(ev, f'今日魔力已经用完，请明天再来~', at_sender=True)
        return
    if not interval.check(uid):
        await bot.finish(ev, f'魔力回复中！(剩余 {int(interval.left_time(uid)) + 1}秒)', at_sender=True)
        return
    daily_limit.increase(uid) 
    interval.start_cd(uid)
    match = ev['match']
    rowid=int(match.group(1))
    await bot.send(ev, f"\n正在炼金中, 请稍后...\n(今日剩余{daily_limit_ - int(daily_limit.get_num(uid))}次)", at_sender=True)

    response = alchemy_manual.use_recipe(rowid)

    if response[0] == 1:
        msg = response[1]
        data = await gen_pic(msg)
        await bot.send_group_forward_msg(group_id=ev["group_id"], messages=data)



@sv.on_rex((r'^删除配方([1-9]\d*)'))
async def delete_recipe(bot, ev):
    if not priv.check_priv(ev, priv.SUPERUSER):
        await bot.finish(ev, '删除配方仅限超级管理员使用', at_sender=True)
        return
    match = ev['match']
    rowid=int(match.group(1))

    response = alchemy_manual.delete_recipe_by_rowid(rowid)
    if response[0] == 1:
        await bot.send(ev, response[1])


async def gen_pic(text):
    try:
        # 去除掉换行符以及空格
        text = text.replace('\n', '').replace(' ', '')
        print(f'text:{text}, type: {type(text)}')
        get_url = f'{word2img_url}?token={token}&tags={text}'
        print(f'正在向{get_url}请求图片')
        res = await aiorequests.get(get_url)
        image = await res.content
        load_data = json.loads(re.findall('{"steps".+?}', str(image))[0])
        image_b64 = 'base64://' + str(base64.b64encode(image).decode())
        # mes = f"[CQ:image,file={image_b64}]"
        # mes += f'\nseed:{load_data["seed"]}'
        data = {"type": "node", "data": {"name": "ai绘图", "uin": "2854196310", "content": f"[CQ:image,file={image_b64}]"}}
        return data
    except Exception as e:
        print(f"炼金失败, 错误信息: {e}")
        return f"炼金失败了,原因:{e}"


ai_draw_group_list = config['ai_draw_group_list']
all_group_list = config['all_group_list']
black_list = []


@sv.on_prefix(('aidraw'))
async def gen_pic_safe(bot, ev: CQEvent):
    try:
        print(f'群号为:{ev.group_id}')
        if ev.group_id not in ai_draw_group_list:
                pass
        else:
            await bot.send(ev, f"正在生成", at_sender=True)
            text = ev.message.extract_plain_text()
            taglist = [chr.lower() for chr in text.split(',')]
            # 遍历 taglist, 如果在 black_list 中则删除
            for i in taglist:
                if i in black_list:
                    taglist.remove(i)
            print(taglist)
            uid = ev.user_id
            for tag in taglist:
                xp.add_xp_num(uid, tag)
            tags = ','.join(str(i) for i in taglist)
            get_url = word2img_url + "?r18=0&tags=" + tags + f'&token={token}'
            print(f'正在请求{get_url}')
            # image = await aiorequests.get(get_url)
            res = await aiorequests.get(get_url)
            image = await res.content
            load_data = json.loads(re.findall('{"steps".+?}', str(image))[0])
            image_b64 = f"base64://{str(base64.b64encode(image).decode())}"
            mes = f"[CQ:image,file={image_b64}]\n"
            mes += f'seed:{load_data["seed"]}   '
            mes += f'scale:{load_data["scale"]}\n'
            mes += f'tags:{text}'
            data = {
                "type": "node", 
                "data": {
                    "name": "QQ小冰", 
                    "uin": "2854196310", 
                    "content": mes
                }
            }
            await bot.send_group_forward_msg(group_id=ev["group_id"], messages=data)

    except:
        await bot.send(ev, "生成失败…")


@sv.on_prefix(('aigener'))
async def gen_pic_all(bot, ev: CQEvent):
    try:
        print(f'群号为:{ev.group_id}')
        if ev.group_id not in all_group_list:
            pass
        else:
            await bot.send(ev, f"正在生成", at_sender=True)
            text = ev.message.extract_plain_text()
            get_url = word2img_url + "?r18=1&tags=" + text + f'&token={token}'
            print(f'正在请求{get_url}')
            # image = await aiorequests.get(get_url)
            res = await aiorequests.get(get_url)
            image = await res.content
            load_data = json.loads(re.findall('{"steps".+?}', str(image))[0])
            image_b64 = 'base64://' + str(base64.b64encode(image).decode())
            mes = f"[CQ:image,file={image_b64}]\n"
            mes += f'seed:{load_data["seed"]}   '
            mes += f'scale:{load_data["scale"]}\n'
            mes += f'tags:{text}'
            await bot.send(ev, mes, at_sender=True)
    except:
        await bot.send(ev, "生成失败…")



@sv.on_prefix("img2img")
async def img2img(bot, ev):
    if ev.group_id not in ai_draw_group_list:
        pass
    else:
        tag = ev.message.extract_plain_text()
        if tag == "":
            url = ev.message[0]["data"]["url"]
        else:
            url = ev.message[1]["data"]["url"]
        await bot.send(ev, "正在生成", at_sender=True)
        image = Image.open(BytesIO(requests.get(url, timeout=20).content))
        # img_x, img_y = int(image.size[0] * (768 / image.size[1])), 768
        # image = image.resize((img_x, img_y))
        thumbSize = (768, 768)
        image = image.convert("RGB")
        if (image.size[0] > image.size[1]):
            image_shape = "Landscape"
        elif (image.size[0] == image.size[1]):
            image_shape = "Square"
        else:
            image_shape = "Portrait"

        image.thumbnail(thumbSize, resample=Image.ANTIALIAS)
        b_io = BytesIO()
        image.save(b_io, format="JPEG")
        posturl =  img2img_url + (f"?tags={tag}&token={token}" if tag != "" else f"?token={token}") 
        resp = await aiorequests.post(
            posturl,
            data=base64.b64encode(b_io.getvalue()),
        )
        print(f'正在向{posturl}发起请求')
        img = await resp.content
        # print(f'返回结果:{img}')
        image_b64 = f"base64://{str(base64.b64encode(img).decode())}"
        data = {"type": "node", "data": {"name": "ai绘图", "uin": "2854196310", "content": f"[CQ:image,file={image_b64}]"}}
        await bot.send_group_forward_msg(group_id=ev["group_id"], messages=data)
